### PR TITLE
[SPARK-19669][SQL] Open up visibility for sharedState, sessionState, and a few other functions

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?
To ease debugging, most of Spark SQL internals have public level visibility. Two of the most important internal states, sharedState and sessionState, however, are package private. It would make more sense to open these up as well with clear documentation that they are internal.

In addition, users currently have way to set active/default SparkSession, but no way to actually get them back. We should open those up as well.

## How was this patch tested?
N/A - only visibility change.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -95,16 +95,26 @@ class SparkSession private(
...
    * State isolated across sessions, including SQL configurations, temporary tables, registered
    * functions, and everything else that accepts a [[org.apache.spark.sql.internal.SQLConf]].
+   *
+   * This is internal to Spark and there is no guarantee on interface stability.
+   *
+   * @since 2.2.0
    */
+  @InterfaceStability.Unstable
   @transient
   private[sql] lazy val sessionState: SessionState = {
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: Like `sharedState`, also remove `private[sql]`?

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: oops

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by windpiger** [comment]
@gatorsmile @rxin 
the latest success test build at commit a0c22a8 not the latest commit 7d3cdd6. 
The open session state will cause the TestSQLContext compile failed, I have make a hotfix for this https://github.com/apache/spark/pull/17008

**Comment by gatorsmile** [comment]
uh... I missed that.

**Comment by rxin** [comment]
Yea @gatorsmile be careful in the future and check the commit hash.

**Comment by gatorsmile** [comment]
Yeah. Will be more careful in the future. Sorry for that.

**Pull request closed without merging**